language/move-model/src/ast.rs

```diff
@@ -354,28 +358,54 @@ pub enum Exp {
     /// (including operators, constants, ...) as well as user functions.
     Call(NodeId, Operation, Vec<Exp>),
     /// Represents an invocation of a function value, as a lambda.
-    Invoke(NodeId, Box<Exp>, Vec<Exp>),
+    Invoke(NodeId, Exp, Vec<Exp>),
     /// Represents a lambda.
-    Lambda(NodeId, Vec<LocalVarDecl>, Box<Exp>),
+    Lambda(NodeId, Vec<LocalVarDecl>, Exp),
     /// Represents a quantified formula over multiple variables and ranges.
     Quant(
         NodeId,
         QuantKind,
+        /// Ranges
         Vec<(LocalVarDecl, Exp)>,
+        /// Triggers
         Vec<Vec<Exp>>,
-        Option<Box<Exp>>,
-        Box<Exp>,
+        /// Optional `where` clause
+        Option<Exp>,
+        // Body
+        Exp,
     ),
     /// Represents a block which contains a set of variable bindings and an expression
     /// for which those are defined.
-    Block(NodeId, Vec<LocalVarDecl>, Box<Exp>),
+    Block(NodeId, Vec<LocalVarDecl>, Exp),
     /// Represents a conditional.
-    IfElse(NodeId, Box<Exp>, Box<Exp>, Box<Exp>),
+    IfElse(NodeId, Exp, Exp, Exp),
 }
 
-impl Exp {
+/// A reference counted expression. Used for sub-expressions to avoid unnecessary copies on
+/// rewriting.
+pub type Exp = ArcIntern<Exp_>;
+
+impl From<Exp> for Exp_ {
+    /// Takes a reference counted expression and returns an owned expression. If the rc is
+    /// exclusive, this will move the underlying value, otherwise a copy will be created.
+    fn from(exp: Exp) -> Exp_ {
+        exp.as_ref().to_owned()
+    }
+}
+
+impl Exp_ {
+    pub fn into_exp(self) -> Exp {
```

Comment:
Not sure if `impl Into<Exp> for Exp_` will help simply many of the changes in this PR due to suffixing a `.into_exp()` to the expression variable.

Reply:
Unfortunately, with the Into trait, in many cases type inference doesn't work, so you have to write `e.into::<Exp>()`. That's why I created the shortcut `into_exp`.